## 1. Installing packages needed

In [2]:
!pip install mne
!pip install yasa

You should consider upgrading via the '/Users/richardpang/opt/anaconda3/bin/python -m pip install --upgrade pip' command.
You should consider upgrading via the '/Users/richardpang/opt/anaconda3/bin/python -m pip install --upgrade pip' command.


In [1]:
import os
import numpy as np
import pandas as pd
import mne
import matplotlib.pyplot as plt
import xml.etree.ElementTree
from xml.etree.ElementTree import parse
from scipy import signal
from scipy.signal import welch
import seaborn as sns
import yasa
%matplotlib qt

## 2.1 Data Loading and Visualizing
### The raw data is an edf file downloaded from the EEG Motor Movement/Imagery Dataset. You may find the dataset here: https://www.physionet.org/content/eegmmidb/1.0.0/. Noted that the data contains 64 channels and the sampling frequency is 160 Hz, as indicated below. 

In [2]:
raw = mne.io.read_raw_edf('data/S001R01.edf')
raw.pick_types(eeg=True)
raw.plot(start=20,duration=1)
plt.show()

Extracting EDF parameters from /Users/richardpang/Desktop/eeg_analysis/data/S001R01.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Using matplotlib as 2D backend.
Opening raw-browser...
Closing raw-browser...
Channels marked as bad:
none


### The plot will come out in a separate window, so I make a screenshot and attach it here for convenience. The actual plot is interactive, not just an image. 

<img src="data/raw_data.png" alt="drawing" width="400"/>

## 2.2 Information about the raw data

In [3]:
# Extract the data and convert from V to uV
data = raw.get_data(units="uV")
sf = raw.info['sfreq']
chan = raw.ch_names

# Let's have a look at the data
print('Chan =', chan)
print('Sampling frequency =', sf, 'Hz')
print('Data shape =', data.shape)

Chan = ['Fc5.', 'Fc3.', 'Fc1.', 'Fcz.', 'Fc2.', 'Fc4.', 'Fc6.', 'C5..', 'C3..', 'C1..', 'Cz..', 'C2..', 'C4..', 'C6..', 'Cp5.', 'Cp3.', 'Cp1.', 'Cpz.', 'Cp2.', 'Cp4.', 'Cp6.', 'Fp1.', 'Fpz.', 'Fp2.', 'Af7.', 'Af3.', 'Afz.', 'Af4.', 'Af8.', 'F7..', 'F5..', 'F3..', 'F1..', 'Fz..', 'F2..', 'F4..', 'F6..', 'F8..', 'Ft7.', 'Ft8.', 'T7..', 'T8..', 'T9..', 'T10.', 'Tp7.', 'Tp8.', 'P7..', 'P5..', 'P3..', 'P1..', 'Pz..', 'P2..', 'P4..', 'P6..', 'P8..', 'Po7.', 'Po3.', 'Poz.', 'Po4.', 'Po8.', 'O1..', 'Oz..', 'O2..', 'Iz..']
Sampling frequency = 160.0 Hz
Data shape = (64, 9760)


In [4]:
print('The actual data is just a matrix array!\n\n {}\n'.format(raw.get_data()))

The actual data is just a matrix array!

 [[-1.6e-05 -5.6e-05 -5.5e-05 ...  0.0e+00  0.0e+00  0.0e+00]
 [-2.9e-05 -5.4e-05 -5.5e-05 ...  0.0e+00  0.0e+00  0.0e+00]
 [ 2.0e-06 -2.7e-05 -2.9e-05 ...  0.0e+00  0.0e+00  0.0e+00]
 ...
 [-2.1e-05 -1.2e-05  2.0e-06 ...  0.0e+00  0.0e+00  0.0e+00]
 [-1.1e-05  1.0e-06  1.8e-05 ...  0.0e+00  0.0e+00  0.0e+00]
 [ 1.5e-05  2.1e-05  3.5e-05 ...  0.0e+00  0.0e+00  0.0e+00]]



In [5]:
raw.info

<Info | 7 non-empty values
 bads: []
 ch_names: Fc5., Fc3., Fc1., Fcz., Fc2., Fc4., Fc6., C5.., C3.., C1.., ...
 chs: 64 EEG
 custom_ref_applied: False
 highpass: 0.0 Hz
 lowpass: 80.0 Hz
 meas_date: 2009-08-12 16:15:00 UTC
 nchan: 64
 projs: []
 sfreq: 160.0 Hz
>

## 3. Plot the PSD of one channel using the welch function

### 3.1 The window size (win) will affect how much detail is included in the PSD. A larger win will reveal more specificity, while a smaller win will make for less variations

In [6]:
win = int(4 * sf)  # Window size is set to 4 seconds
freqs, psd = welch(data, sf, nperseg=win, average='median')  # Works with single or multi-channel data
# freqs, psd = welch(data, sf, average='median')  # Works with single or multi-channel data

print(freqs.shape, psd.shape)  # psd has shape (n_channels, n_frequencies)

# Plot
plt.plot(freqs, psd[1], 'k', lw=2)
plt.fill_between(freqs, psd[1], cmap='Spectral')
plt.xlim(0, 80)
plt.yscale('log')
sns.despine()
plt.title(chan[1])
plt.xlabel('Frequency [Hz]')
plt.ylabel('PSD log($uV^2$/Hz)');

(321,) (64, 321)


The PSD with a win of 1 * 160 = 160:

 <img src="data/win_160.png" alt="drawing" width="400"/>

The PSD with a win of 4 * 160 = 640:

 <img src="data/win_640.png" alt="drawing" width="400"/>

### 3.2 The calculation of bands power are largely dependent on the window size as well. As indicated in the documentation of yasa library, I use win = int(4 * sf) = 640 in this numerical analysis. 

## 4.1 Calculate band powers from PSD (relative and absolute)

#### Here are the range of each band, we need to make sure the range of Delta, Theta, etc. are consistent when comparing results obtained here with the results obtained somewhere else. The range of each band is shown below.

In [7]:
bands=[0.5, 4, 'Delta', 4, 8, 'Theta', 8, 12, 'Alpha', 12, 16, 'Sigma', 16, 30, 'Beta', 30, 40, 'Gamma']

#### Yasa is a neuroscience package developed from UC Berkley. After doing some research, I believe this is the easiest way to compute the band power. <br> </br> Relative band power refers to the ratio of these bands in the raw data, absolute band power stands for the absolute value, in which the unit is uV^2/Hz.

In [8]:
yasa.bandpower_from_psd(psd, freqs, ch_names=chan)

Chan     Delta     Theta     Alpha     Sigma      Beta     Gamma  \
0   Fc5.  0.546720  0.188663  0.081719  0.071354  0.086720  0.024824   
1   Fc3.  0.553381  0.186151  0.088004  0.079254  0.075679  0.017531   
2   Fc1.  0.577314  0.180922  0.088535  0.069158  0.068330  0.015740   
3   Fcz.  0.625901  0.160527  0.073067  0.059309  0.067279  0.013917   
4   Fc2.  0.638252  0.153374  0.070831  0.061820  0.061709  0.014013   
..   ...       ...       ...       ...       ...       ...       ...   
59  Po8.  0.611308  0.128695  0.089872  0.074769  0.075307  0.020049   
60  O1..  0.551645  0.133374  0.107755  0.091970  0.102906  0.012351   
61  Oz..  0.590588  0.124811  0.089765  0.078538  0.100867  0.015432   
62  O2..  0.616492  0.120480  0.085481  0.067933  0.093260  0.016354   
63  Iz..  0.518119  0.134041  0.100395  0.080367  0.134792  0.032286   

    TotalAbsPow  FreqRes  Relative  
0   1333.560137     0.25      True  
1   1591.182522     0.25      True  
2   1711.810271     0.25      True  
3   1837.897807     0.25      True  
4   1706.774518     0.25      True  
..          ...      ...       ...  
59  1463.046097     0.25      True  
60  1852.322341     0.25      True  
61  1840.485750     0.25      True  
62  1992.397458     0.25      True  
63  1728.998201     0.25      True  

[64 rows x 10 columns]

In [9]:
yasa.bandpower_from_psd(psd, freqs, ch_names=chan,relative=False)

Chan        Delta       Theta       Alpha       Sigma        Beta  \
0   Fc5.   729.083449  251.593684  108.977588   95.155092  115.646269   
1   Fc3.   880.530178  296.200884  140.030286  126.106843  120.418670   
2   Fc1.   988.252429  309.703739  151.555804  118.386149  116.968244   
3   Fcz.  1150.341449  295.031470  134.289831  109.004222  123.652382   
4   Fc2.  1089.352459  261.774983  120.893042  105.512753  105.323430   
..   ...          ...         ...         ...         ...         ...   
59  Po8.   894.372006  188.287211  131.487208  109.390743  110.176915   
60  O1..  1021.823602  247.050760  199.597288  170.357692  190.614198   
61  Oz..  1086.968179  229.712791  165.211522  144.547493  185.643865   
62  O2..  1228.297022  240.044884  170.311805  135.349890  185.810987   
63  Iz..   895.827385  231.756621  173.582004  138.955198  233.054949   

        Gamma  TotalAbsPow  FreqRes  Relative  
0   33.104056  1333.560137     0.25     False  
1   27.895661  1591.182522     0.25     False  
2   26.943906  1711.810271     0.25     False  
3   25.578453  1837.897807     0.25     False  
4   23.917851  1706.774518     0.25     False  
..        ...          ...      ...       ...  
59  29.332015  1463.046097     0.25     False  
60  22.878800  1852.322341     0.25     False  
61  28.401902  1840.485750     0.25     False  
62  32.582869  1992.397458     0.25     False  
63  55.822044  1728.998201     0.25     False  

[64 rows x 10 columns]

## 4.2 Calculate band powers from raw data (relative and absolute)

In [10]:
# Relative bandpower per channel on the whole recording (entire data)
yasa.bandpower(raw, sf=sf, ch_names=chan)

Delta     Theta     Alpha     Sigma      Beta     Gamma  TotalAbsPow  \
Chan                                                                            
Fc5.  0.548339  0.188002  0.085510  0.069651  0.084315  0.024183  1356.084169   
Fc3.  0.569570  0.174826  0.087572  0.076244  0.074501  0.017287  1627.558176   
Fc1.  0.581620  0.177363  0.086579  0.070425  0.068284  0.015729  1695.851253   
Fcz.  0.611424  0.169222  0.074378  0.062707  0.068008  0.014262  1821.866101   
Fc2.  0.634056  0.153610  0.074786  0.061335  0.061986  0.014227  1689.180796   
...        ...       ...       ...       ...       ...       ...          ...   
Po8.  0.611426  0.127151  0.089966  0.074720  0.076852  0.019886  1461.909318   
O1..  0.555521  0.130322  0.107836  0.090734  0.103069  0.012518  1858.856100   
Oz..  0.588168  0.122236  0.092429  0.078167  0.103260  0.015741  1828.575233   
O2..  0.617893  0.118152  0.085242  0.068090  0.093996  0.016627  2017.668395   
Iz..  0.517579  0.133765  0.101682  0.081009  0.133400  0.032564  1728.198818   

      FreqRes  Relative  
Chan                     
Fc5.     0.25      True  
Fc3.     0.25      True  
Fc1.     0.25      True  
Fcz.     0.25      True  
Fc2.     0.25      True  
...       ...       ...  
Po8.     0.25      True  
O1..     0.25      True  
Oz..     0.25      True  
O2..     0.25      True  
Iz..     0.25      True  

[64 rows x 9 columns]

In [11]:
yasa.bandpower(raw,relative=False)

Delta       Theta       Alpha       Sigma        Beta      Gamma  \
Chan                                                                           
Fc5.   743.594341  254.945948  115.958546   94.453281  114.337948  32.794104   
Fc3.   927.008354  284.540254  142.528072  124.091109  121.254069  28.136317   
Fc1.   986.341575  300.781044  146.824816  119.430047  115.800208  26.673562   
Fcz.  1113.932683  308.299800  135.506160  114.242997  123.900846  25.983616   
Fc2.  1071.034496  259.475381  126.327908  103.606599  104.704983  24.031429   
...           ...         ...         ...         ...         ...        ...   
Po8.   893.849195  185.882762  131.521556  109.234149  112.350669  29.070987   
O1..  1032.634114  242.250511  200.450749  168.661942  191.589984  23.268800   
Oz..  1075.508674  223.517801  169.012913  142.934222  188.818328  28.783294   
O2..  1246.703368  238.391781  171.990011  137.383224  189.652426  33.547585   
Iz..   894.479945  231.173025  175.726501  140.000437  230.541963  56.276947   

      TotalAbsPow  FreqRes  Relative  
Chan                                  
Fc5.  1356.084169     0.25     False  
Fc3.  1627.558176     0.25     False  
Fc1.  1695.851253     0.25     False  
Fcz.  1821.866101     0.25     False  
Fc2.  1689.180796     0.25     False  
...           ...      ...       ...  
Po8.  1461.909318     0.25     False  
O1..  1858.856100     0.25     False  
Oz..  1828.575233     0.25     False  
O2..  2017.668395     0.25     False  
Iz..  1728.198818     0.25     False  

[64 rows x 9 columns]

### 4.3 There are slight differences between the results calculated from PSD and the raw data. I think it may be because of a tiny loss from the PSD calculation process, since the plot cannot be as concrete as the raw data. <br> </br> Also, the results obtained from different approaches (e.g. the welch approach and the Multitaper approach) may result in similar but not identical results. 

## 5. Calculate band powers from Matlab (eeglab)

### 5.1 Find the PSD of the i-th signal, 9760 is the length of data

```
[spectra,freqs] = spectopo(EEG.data(i,:,:), 9760, EEG.srate)
```

### 5.2 Define the range of each band, same as before

```
deltaIdx = find(freqs>0.5 & freqs<4);
thetaIdx = find(freqs>4 & freqs<8);
alphaIdx = find(freqs>8 & freqs<12);
sigmaIdx = find(freqs>12 & freqs<16);
betaIdx  = find(freqs>16 & freqs<30);
gammaIdx = find(freqs>30 & freqs<40); 
```

### 5.3 The function of calculating band power from eeglab

```
deltaPower = mean(10.^(spectra(deltaIdx)/10));
thetaPower = mean(10.^(spectra(thetaIdx)/10));
alphaPower = mean(10.^(spectra(alphaIdx)/10));
sigmaPower = mean(10.^(spectra(sigmaIdx)/10));
betaPower  = mean(10.^(spectra(betaIdx)/10));
gammaPower = mean(10.^(spectra(gammaIdx)/10));
```

### 5.4 Display Results 

In [36]:
import pandas as pd
pd.read_csv('data/bandpower_matlab.csv')

Delta      Theta      Alpha      Sigma       Beta     Gamma
0   406.566768  63.870613  25.694729  21.686916   7.993780  3.273779
1   403.217832  74.586141  28.853016  26.432134   8.383048  2.701803
2   417.797976  80.677422  27.029659  25.151542   8.330458  2.489914
3   402.369663  84.618613  26.406469  26.478589   8.489373  2.411848
4   397.411402  72.139138  24.121785  26.030127   7.562521  2.276452
..         ...        ...        ...        ...        ...       ...
59  281.572576  39.654170  29.859911  30.463295   8.462207  2.791146
60  320.641388  54.802062  42.525765  38.810288  13.473526  2.346743
61  325.898590  51.139658  37.262216  34.790947  12.617842  2.798073
62  456.382881  54.744618  38.175265  37.230773  13.606381  3.296833
63  319.314642  49.876303  37.602395  31.276189  16.268192  5.235330

[64 rows x 6 columns]

### The results are different from what I obtained before. Many factors could lead to this inconsistency, such as window size, calculatipon function, etc., I am not sure why they are different since I am not an expert in neuroscience. 

## 6. Anomalies in eeg 

#### I googled the term "anomalies in eeg" and it seems like anomaly detection is a difficult problem which requires a lot of steps including some complex architectures and even some machine learning approaches. It may take me weeks if I really want to do some research and carefully study this field. For this reason, I decide not to do this part for now.